lib/log_sampler.js

```diff
@@ -0,0 +1,59 @@
+"use strict";
+
+/**
+ * Creates a log sampler. In case the log rate is not exceeded, all the logs
+ * are being written. Otherwise starts sampling the logs with decreasing
+ * probability.
+ *
+ * @param {object} options:
+ * @param {Number} [options.limit] The maximum rate of sampled logs per interval. Default: 1000.
+ * @param {Number} [options.interval] Update interval in ms. Default: 1000ms.
+ * @param {Number} [options.minValue] Drop global counters below this value. Default: 10.
+ */
+function Sampler(options) {
+    this._options = options === true ? {} : options || {};
+    this._options.limit = this._options.limit || 1000;
+    this._options.interval = this._options.interval || 1000;
+    this._options.minValue = this._options.minValue || 10;
+    // TODO: replace with a native Map when support for node 0.10 is dropped.
+    this._counters = {};
+    // Exponential decay with factor 2, thrice per interval for better
+    // smoothness, using the cubic root of 2).
+    this._decayFactor = Math.pow(2, 1 / 3);
+    this._decayedLimit = this._options.limit / this._decayFactor;
+    this._decayInterval = setInterval(this._decay.bind(this), this._options.interval / 3);
```

Comment:
If you calculate the sampling rate right after decay, then you don't need to schedule extra updates, as the only time the rate will change would be once per interval.